WD-2175 Career progression

carkod opened the pull request:
## Done

Build career/progression page
## QA

- View https://canonical-com-823.demos.haus/careers/progression and compare with [design](https://www.figma.com/file/UsovUAf4zELQgGd9bxhiVe/Careers?node-id=1496-11228&t=ZOwzdx88YOIY8Rnm-0)
- Check everything looks good on mobile
- Check the other careers search pages work (this PR has refactored search) e.g. https://canonical-com-823.demos.haus/careers/all and https://canonical-com-823.demos.haus/careers. Do a quick search
- Run through the following [QA steps](https://canonical-web-and-design.github.io/practices/workflow/qa-steps.html)
- [List additional steps to QA the new features or prove the bug has been resolved]

## Fun fact/Behind-the-scenes

[About 6 codepens](https://codepen.io/carkod/pens/public?cursor=ZD0wJm89MSZwPTEmdj02NzE0ODcyNQ==) were created before putting together this page.


## Issue / Card

Fixes 

## Screenshots

[if relevant, include a screenshot]

webteam-app commented:
Demo starting at https://canonical-com-823.demos.haus

anthonydillon commented on static/js/careers-progression.js:
```diff
@@ -0,0 +1,272 @@
+  }
+  
+  
+  function shift_to_right(elem) {
```
Function names should be camelcase

anthonydillon reviewed (changes requested):
Wow @carkod this looks great. I don't think you are too far off. I left some comments in the code and there are some improvements to the small screen view.

carkod commented:
> Hi @carkod Thank you for building this so fast. It looks cool! I do have feedback smile
> 
> ![Screen Shot 2023-03-23 at 1 13 30 PM](https://user-images.githubusercontent.com/112855834/227177634-ded16d38-f6d2-4fbd-b755-a94906c904d3.png)
> 
> * the line should be inside the column with 22px padding-right.
> * the line opacity is a 0.25
> * there's a m,issing line on the leftmost content.
> 
> ![Screen Shot 2023-03-23 at 1 14 08 PM](https://user-images.githubusercontent.com/112855834/227174382-374a6221-9354-4356-ac9f-c79a51012472.png)
> 
> * Is it possible that the photos genuinely shuffle back and smoothly go into a loop with the same content?
> * The text isn't changing as you click through
> 
> <img alt="Screen Shot 2023-03-23 at 1 14 55 PM" width="1446" src="https://user-images.githubusercontent.com/112855834/227174826-b066befb-15e3-4c1a-b540-8af6e1ee2e04.png">
> 
> - add break
> ```
> <div><h#> To deliver the world's best open source <br> experience, from platform to application<h#><div>
> ```
> 
> * remove break on small screen sizes.
> 
> <img alt="Screen Shot 2023-03-23 at 1 15 05 PM" width="1446" src="https://user-images.githubusercontent.com/112855834/227175141-3553abaf-2130-4db5-9af7-1dbb7cbcb49b.png">
> 
> - the background color of the card has some transparency just make it a solid #fff 
> Thank you! I'll make a mobile version view of the whole thing. So it translates better in smaller screen sizes. I found issues with the mobile view that can be explained better visually. Again thank you for your hard work on this, Carlos. I genuinely appreciate it.

Addressed all comments!

carkod commented on static/sass/_pattern_careers-progression.scss:
```diff
@@ -0,0 +1,360 @@
+    }
+  }
+
+  // Custom animations
+  // @keyframes slide-right {
+  // 	0% {
+  // 		opacity: 0;
+  // 		transform: translate3d(-50%, -20%, 0);
+  // 	}
+  // 	50% {
+  // 		opacity: 0;
+  // 		transform: translate3d(-25%, -10%, 0);
+  // 	}
+  // 	100% {
+  // 		opacity: 1;
+  // 		transform: translate3d(0, 0, 0);
+  // 	}
+  // }
+
+  // .u-animation--slide-right {
+  // white-space: nowrap;
+  // display: block;
+  // overflow: hidden;
+  // margin: 3rem 0;
+  // animation: type 4s 1 forwards linear;
+  // }
```
I left these comments on purpose in case design wants to change some of the animations. Once I get design review +1 I'll remove them.

carkod replied:
This has been removed now.

carkod commented on static/sass/_pattern_careers-progression.scss:
```diff
@@ -0,0 +1,360 @@
+      }
+    }
+
+    .p-roles {
```
Removing them causes the images not to align and design has very specific spacing that is not even and there is a border on the left, all those styles are covered by this class.

carkod commented on static/sass/_pattern_careers-progression.scss:
```diff
@@ -0,0 +1,360 @@
+    }
+
+    // Reset Vanilla styles
+    video {
```
This is overriding the current Vanilla styles which cause the element to align in the center and some other additional padding, it's not a utility to add styling it's correction. Vanilla base styles add styles to the `video` element by defeault

carkod commented on static/sass/_pattern_careers-progression.scss:
```diff
@@ -0,0 +1,360 @@
+@import "../../node_modules/vanilla-framework/scss/patterns_icons";
+
+@mixin canonical-p-careers-progression {
+  .p-strip--careers-progression {
```
I can use the function, but the color is obviously not the same.

carkod commented on static/sass/_pattern_careers-progression.scss:
```diff
@@ -0,0 +1,360 @@
+      justify-content: space-between;
+    }
+
+    .p-strip--hero-title {
```
That is not intended for reuse. This hero section has anitmation that other pages don't have, and has a lot more padding and spacing than other careers pages. I could append the `careers-progression` prefix, but the class name was getting very long...

marinacastejon commented:
Hi @carkod I found a few things:
- The images are not the same height
<img width="1156" alt="image" src="https://user-images.githubusercontent.com/89836430/228185639-3cd12926-10bf-4b33-b60d-adb0a4450d8c.png">
- The chevrons are not center-aligned in the 2 sections where they appear
<img width="126" alt="image" src="https://user-images.githubusercontent.com/89836430/228186235-9f0271db-21de-4613-8e8e-ecdf34a665ab.png">
- The video is not accessible the way it is, I'm leaving a comment for @camillet16  in Figma proposing accessible alternatives. 
Thanks, 
Marina

anthonydillon commented on static/js/careers-progression.js:
```diff
@@ -0,0 +1,272 @@
+  // https://github.com/canonical/canonical.com/blob/main/templates/index.html#L587
+  const elementsToAnimate = document.querySelectorAll('[data-animation]');
+  let windowHeight = window.innerHeight;
+  window.addEventListener("scroll", () => {
```
Did you try this @carkod?

carkod replied:
Yes, because this needs to add an observer to every element, that needs to animate, it doesn't seem to work.

carkod replied:
Why not? This is what's being used on the homepage?

anthonydillon replied:
s/wouldn't/would

anthonydillon commented on static/js/careers-progression.js:
```diff
@@ -0,0 +1,272 @@
+    elementsToAnimate.forEach(element => {
+      var positionFromTop = element.getBoundingClientRect().top;
+      if (positionFromTop - windowHeight <= 0) {
+        element.classList.add(element.dataset.animation);
```
Did you try this @carkod?

anthonydillon replied:
This is setting an undefined class

carkod replied:
Fixed

anthonydillon commented on static/js/careers-progression.js:
```diff
@@ -0,0 +1,298 @@
+
+function setupEdgeHugger() {
+  /**
+   * Layout to huge the right edge of the page
```
```suggestion
   * Layout to hug the right edge of the page
```

anthonydillon commented on static/js/careers-progression.js:
```diff
@@ -0,0 +1,298 @@
+  }
+}
+
+function setupEdgeHugger() {
```
I believe this would be more performant as CSS. Example: https://codepen.io/anthonydillon/pen/JjaVYJR?editors=1100

anthonydillon commented on static/sass/_pattern_careers-progression.scss:
```diff
@@ -0,0 +1,378 @@
+      --width: 110px;
+    }
+
+    animation: slide-right map-get($animation-duration, slow) 1 forwards linear;
```
The linear easing here feels very robotic.

carkod replied:
Added more custom `cubic-benzier` but let me know if you want something more specific https://cubic-bezier.com/

anthonydillon reviewed (changes requested):
Few more comments there @carkod. Let me know if you want to discuss any.

carkod commented:
@marinacastejon @camillet16 can I get another review? Thanks

carkod commented:
> @carkod @camillet16 After reading Carlos' comments in figma, I think it might be worth re-prioritising things and postponing the video. If to make it accessible it's going to be a lot of work having the video, then I suggest launching this only with an image and creating a separate issue just to work on the accessible video. I think there are other things with a higher priority in this project, such as the meganav. What do you think? FYI @anthonydillon

Well, adding controls is ok. It's already done, if you check the demo. But maybe video accessibility is more than just adding controls?

anthonydillon commented:
@carkod The video looks good to me. The only a11y guideline not hit is the ability to hide the video but I think the pause is good for now.

carkod commented on static/sass/_pattern_careers-progression.scss:
```diff
@@ -0,0 +1,360 @@
+    color: #3097ff;
+  }
+
+  .u-sr-only {
```
Removed

carkod commented on static/sass/_pattern_careers-progression.scss:
```diff
@@ -0,0 +1,360 @@
+
+    animation: slide-right 0.75s 1 forwards linear;
+    display: block;
+    margin: 3rem 0;
```
Thanks. Moved to the title class

anthonydillon commented on templates/careers/progression.html:
```diff
@@ -1,32 +1,343 @@
+  <div class="row">
+    <div class="col-4">
+      <h2>We value people who love to learn</h2>
+      <p>not just how to be better engineers but how to be better speakers, how to be better designers, better organizers, better partners, and better vendors.</p>
```
```suggestion
      <p>Not just how to be better engineers but how to be better speakers, how to be better designers, better organisers, better partners, and better vendors.</p>
```

anthonydillon commented on templates/careers/progression.html:
```diff
@@ -1,32 +1,343 @@
+      </div>
+    </div>
+    <div class="col-4 col-medium-3" data-animation="u-animation--slide-from-bottom" style="animation-delay: 0.50s">
+      <div class="p-roles" style="border-right: 0;">
```
This online style can be more elegantly handled in CSS `last-of-type`

carkod replied:
So I think the trade-off is that it will become more messy in the stylesheet. It's not exactly `last-child` it's actually last-child of a different parent i.e. a "cousin". So to use a selector for the cousin means I will need to create a parent selector (new parent selector? Because I'm using Vanilla styles mostly here, no custom parent class) that goes down to the child, just so I can add one line of CSS.

anthonydillon commented on templates/careers/progression.html:
```diff
@@ -1,32 +1,343 @@
+      {{
+        image (
+        url="https://assets.ubuntu.com/v1/646204f1-Elektrobit 1.png",
+        alt="EB corbos Linux with Elektrobit",
```
Does this sentence describe the image to someone that can't see it?

carkod replied:
Well, I can remove it, but the a11y validator complains of missing `alt`

anthonydillon commented on templates/careers/progression.html:
```diff
@@ -1,32 +1,343 @@
+      {{
+        image (
+        url="https://assets.ubuntu.com/v1/058ff4a7-nvidia.png",
+        alt="Ubuntu core",
```
Does this sentence describe the image to someone that can't see it?

anthonydillon commented on templates/careers/progression.html:
```diff
@@ -1,32 +1,343 @@
+    </div>
+    <div class="col-6 col-medium-3">
+      <small>20 September 2023</small>
+      <h3>Ubuntu Core set to redefine industrial computing with new edge AI platform NVIDIA IGX</h3>
```
```suggestion
      <h3>Ubuntu Core is set to redefine industrial computing with new edge AI platform NVIDIA IGX</h3>
```

anthonydillon commented on templates/careers/progression.html:
```diff
@@ -1,32 +1,343 @@
+  <div class="u-fixed-width">
+    <h2>What team members say</h2>
+  </div>
+</section>
```
Could you merge this section with the one below somehow as they are connected content?

anthonydillon commented on templates/careers/progression.html:
```diff
@@ -1,32 +1,343 @@
+  <div class="p-3d-carousel u-hide--small u-hide--medium">       
+    <div class="p-3d-carousel__container">    
+      <div class="p-3d-carousel__item" aria-current="true">
+        <p>“At Canonical, every day is different. I am able to learn a lot. The people are consistently amazing.”</p>
```
It's a shame to not have citations for these quotes. It would make them more personal and real

anthonydillon commented on templates/careers/progression.html:
```diff
@@ -1,32 +1,343 @@
+</section>
+
+
+<section class="u-fixed-width u-no-padding--top p-strip u-hide--small u-hide--medium" style="max-width: 525px; padding-left: 0; padding-right: 0;">
```
Never seen `u-fixed-width` and `p-strip` used together. Was this by design?

anthonydillon commented on templates/careers/progression.html:
```diff
@@ -1,32 +1,343 @@
+  </div>
+</section>
+
+<section class="p-strip u-fixed-width">
```
And again but doubled up?

anthonydillon commented on webapp/app.py:
```diff
@@ -217,6 +217,62 @@ def start_career():
+    get_vacancies = greenhouse.get_vacancies()
+    all_departments = (_group_by_department(get_vacancies),)
+
+    dept_list = [
```
Is this logic now duplicated? If so, we should centralise it with a reuseable util.

carkod replied:
Ok I think I found a way to achieve both.

anthonydillon reviewed (changes requested):
Really good work here @carkod. Some general comments in the code.

anthonydillon commented on templates/careers/progression.html:
```diff
@@ -1,32 +1,343 @@
+      }}
+    </div>
+    <div class="col-6 col-medium-3">
+      <small>20 September 2023</small>
```
```suggestion
      <small>20 September 2022</small>
```
This is the future

anthonydillon reviewed (commented):
The line height on the hero title seems too large on small screens.
![Screenshot_2023-03-31-03-24-44-42_40deb401b9ffe8e1df2f1cc5ba480b12.jpg](https://user-images.githubusercontent.com/1413534/229007070-130eaeb4-e438-4a06-8323-031e4719fc64.jpg)

There is a little x-axis scrolling in the phone.

carkod commented:
> @carkod The video looks good to me. The only a11y guideline not hit is the ability to hide the video but I think the pause is good for now.

The video hides with screen readers via `aria-hidden`

carkod commented:
> The line height on the hero title seems too large on small screens.
> 
> There is a little x-axis scrolling in the phone.

Fixed

carkod commented on webapp/app.py:
```diff
@@ -109,6 +77,116 @@ def _group_by_department(vacancies):
     return vacancies_by_department
 
 
+def _get_sorted_departments():
```
This is just moving the code up, so that all `_` helper functions are in one place to be referenced.

carkod commented:
> * https://canonical-com-823.demos.haus/careers/progression

Fixed

anthonydillon commented on static/js/careers-progression.js:
```diff
@@ -0,0 +1,278 @@
+  const reducedMotion = window.matchMedia('(prefers-reduced-motion: reduce)').matches;
+  if (reducedMotion) {
+    const videoElement = document.querySelectorAll("video");
+    videoElement.removeAutoplay();
```
videoElement here will be an array. You are likely to hit an error.

anthonydillon commented on static/sass/_pattern_careers-progression.scss:
```diff
@@ -0,0 +1,379 @@
+@mixin canonical-p-careers-progression {
+  body {
+    // Fix overflown video
+    overflow-x: hidden;
```
This will affect all body elements on this site. Is that expected? If so, I would suggest putting it somewhere central like styles.scss

anthonydillon commented on static/sass/_pattern_careers-progression.scss:
```diff
@@ -0,0 +1,379 @@
+    }
+  }
+
+  .p-separator {
```
Won't this animate all the separators across the site?

carkod replied:
I couldn't achieve the same effect with the Vanilla styles, so I renamed it as a custom class

anthonydillon commented on tests/test_routes.py:
```diff
@@ -68,6 +68,16 @@ def test_invalid_careers_department(self):
+        """
+
+        self.assertEqual(
+            self.client.get("/careers/progression").status_code, 200
```
👍

anthonydillon reviewed (approved):
LGTM Carlos, few small comments but good to go once addressed 👍
